### ISSUE TITLE
CORS errors in 5.1.3000

### ISSUE THREAD

**Issue by cjstremick** [datalust/seq-tickets#838]
We just updated to 5.1.3000 and our client apps (using structured-log) are now failing to log because Seq is no longer sending CORS headers to the browser.  We have configured "corsAllowOrigins" to include our applications base Uri, which used to be sufficient.

Has something changed related to this?  Is there new configuration to let Seq send CORS headers?

**Comment by nblumhardt** [comment]
Thanks for the heads-up, @cjstremick - yes, some fixes were made in this area. Did you upgrade from 5.0.x, or 5.1.x?

**Comment by KodrAus** [comment]
Hi @cjstremick, sorry about that, is it possible to get a dump of some raw request/response headers for one of your client apps to Seq's `/api/events/raw` endpoint along with the exact error the browser is reporting? If there's an `OPTIONS` request that'll be the one, otherwise it'll be a `POST`.

**Comment by cjstremick** [comment]
> Thanks for the heads-up, @cjstremick - yes, some fixes were made in this area. Did you upgrade from 5.0.x, or 5.1.x?

We went from 5.0.2817 to 5.1.3000

**Comment by cjstremick** [comment]
@KodrAus, I reverted to 5.0.2817 to get things flowing again so I can't check just now.  I'll try to make time to install 5.1.3000 and reproduce the error to get the details tomorrow.

I do know that the `OPTIONS` request didn't have any CORS headers which caused the browser (Firefox, in my case) to not make a subsequent `POST` to `https://devseq.w******s.com/api/events/raw?apiKey=*****`.

We also saw some HTTP 504's in the console which may have been related (and unusual).  I'll try to capture that information as well.

**Comment by cjstremick** [comment]
@KodrAus - Here's some information:

The request headers for the `OPTIONS` request to seq:
```
Host: devseq.w***s.com
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:66.0) Gecko/20100101 Firefox/66.0
Accept: */*
Accept-Language: en-US,en;q=0.5
Accept-Encoding: gzip, deflate, br
Access-Control-Request-Method: POST
Access-Control-Request-Headers: content-type
Referer: https://devc***t.a***s.net/
Origin: https://devc***t.a***s.net
DNT: 1
Connection: keep-alive
TE: Trailers
```

Here are the response headers from the request (5.1.3000):
```
HTTP/2.0 200 OK
content-length: 0
server: Microsoft-HTTPAPI/2.0
x-content-type-options: nosniff
x-xss-protection: 1; mode=block
date: Thu, 11 Apr 2019 12:36:40 GMT
X-Firefox-Spdy: h2
```

And here's what we see in the console...

![seq-cors-error](https://user-images.githubusercontent.com/3650706/55958056-676d2280-5c2d-11e9-9e65-b2e28ff9c703.png)


Here are the response headers from the request (5.0.2817):
```
HTTP/2.0 204 No Content
vary: Origin
server: Microsoft-HTTPAPI/2.0
access-control-allow-origin: https://devc***t.a***s.net
access-control-allow-headers: content-type
x-content-type-options: nosniff
x-xss-protection: 1; mode=block
date: Thu, 11 Apr 2019 12:45:18 GMT
X-Firefox-Spdy: h2
```

So besides the obvious difference in the access-control-allow-origin header (which matches our apps scheme+host), the other difference is that we're getting a 200 rather than a 204 which I'm not sure is significant.

**Comment by KodrAus** [comment]
Thanks for the details @cjstremick Assuming those requests are to `/api/events/raw` then that definitely doesn't look right!

When you've got specifically allowed origins through `api.corsAllowedOrigins` we should have two CORS policies active:

- Any origin is allowed to `/api/events/raw`
- Only specific origins are allowed to other endpoints

The `/api/events/raw` policy should always be active, but if you're not getting any CORS headers back from `OPTIONS` there could be a few different things at work on the server end.

Are you happy to drop us a line at `support@datalust.co` and give a pre-release build a try?

**Comment by Sebbl22** [comment]
I can confirm, that there is some kind of problem. We updated from 5.0.2817 to 5.1.3000 and had to go immediately back since the "any origin to ` /api/events/raw`" policy and the api.corsAllowedOrigins-setting were not working.

Here is the relevant section as Firefox HAR-Export. No subsequent POSTs were sent.
```json
{
        "pageref": "page_1",
        "startedDateTime": "2019-04-16T15:17:56.480+02:00",
        "request": {
          "bodySize": 0,
          "method": "OPTIONS",
          "url": "https://SUB.LOGDOMAIN.lan/api/events/raw?clef",
          "httpVersion": "HTTP/2.0",
          "headers": [
            {
              "name": "Host",
              "value": "SUB.LOGDOMAIN.lan"
            },
            {
              "name": "User-Agent",
              "value": "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:66.0) Gecko/20100101 Firefox/66.0"
            },
            {
              "name": "Accept",
              "value": "*/*"
            },
            {
              "name": "Accept-Language",
              "value": "de,en-US;q=0.7,en;q=0.3"
            },
            {
              "name": "Accept-Encoding",
              "value": "gzip, deflate, br"
            },
            {
              "name": "Access-Control-Request-Method",
              "value": "POST"
            },
            {
              "name": "Access-Control-Request-Headers",
              "value": "content-type,jsnlog-requestid,x-seq-apikey"
            },
            {
              "name": "Referer",
              "value": "https://SUB.SUB.DOMAIN.lan/?FilialNr=1234"
            },
            {
              "name": "Origin",
              "value": "https://SUB.SUB.DOMAIN.lan"
            },
            {
              "name": "Connection",
              "value": "keep-alive"
            },
            {
              "name": "TE",
              "value": "Trailers"
            }
          ],
          "cookies": [],
          "queryString": [
            {
              "name": "clef",
              "value": ""
            }
          ],
          "headersSize": 514
        },
        "response": {
          "status": 200,
          "statusText": "OK",
          "httpVersion": "HTTP/2.0",
          "headers": [
            {
              "name": "content-length",
              "value": "0"
            },
            {
              "name": "server",
              "value": "Microsoft-HTTPAPI/2.0"
            },
            {
              "name": "strict-transport-security",
              "value": "max-age=31536000"
            },
            {
              "name": "x-content-type-options",
              "value": "nosniff"
            },
            {
              "name": "x-xss-protection",
              "value": "1; mode=block"
            },
            {
              "name": "date",
              "value": "Tue, 16 Apr 2019 13:17:56 GMT"
            },
            {
              "name": "X-Firefox-Spdy",
              "value": "h2"
            }
          ],
          "cookies": [],
          "content": {
            "mimeType": "text/plain",
            "size": 0,
            "text": ""
          },
          "redirectURL": "",
          "headersSize": 237,
          "bodySize": 237
        },
        "cache": {},
        "timings": {
          "blocked": 0,
          "dns": 0,
          "connect": 0,
          "ssl": 0,
          "send": 0,
          "wait": 26,
          "receive": 0
        },
        "time": 26,
        "_securityState": "secure",
        "serverIPAddress": "1.2.3.4",
        "connection": "443"
      },```

**Comment by KodrAus** [comment]
Hey @Sebbl22, yes unfortunately it looks like we regressed the CORS behaviour on the ingestion endpoint in 5.1.3000 but have a patch that fixes it up. It'll make its way out soon (we'll keep you posted) but in the meantime, if you do log to Seq from the browser you can contact us at support@datalust.co to give the pre-release build a try or avoid upgrading until we've got that patch out.

**Comment by KodrAus** [comment]
We've just published `5.1.3004` that includes the reworked CORS implementation.

**Comment by Sebbl22** [comment]
Thanks, both CORS-Bugs are now fixed for us.
